src/Microsoft.AspNetCore.Routing/Patterns/DefaultRoutePatternTransformer.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    internal class DefaultRoutePatternTransformer : RoutePatternTransformer
+    {
+        private readonly ParameterPolicyFactory _policyFactory;
+
+        public DefaultRoutePatternTransformer(ParameterPolicyFactory policyFactory)
+        {
+            if (policyFactory == null)
+            {
+                throw new ArgumentNullException(nameof(policyFactory));
+            }
+
+            _policyFactory = policyFactory;
+        }
+
+        public override RoutePattern SubstituteRequiredValues(RoutePattern original, object requiredValues)
```

Comment:
I'm a little confused about this returning a single route pattern. One of the reasons the logic in MVC was so complicated is a route pattern denormalized to multiple endpoints because of defaults.

{controller=Home}/{action=Index}/{id?} became
* /
* /Home
* /Home/Index/{id}

You're returning one pattern but removing the parameters. How will the necessary permutations be calculated?

Reply:
There is no need for permutations - the DFA matcher will do the right thing (once it's implemented). This is a new fundamental feature of `RoutePattern`

Reply:
> You're returning one pattern but removing the parameters

I am not removing the parameters. I am removing a default value when it conflicts with a required value.